Implement kubeadm bootstrap token management

dgoodwin opened the pull request:
Creates bootstrap tokens as secrets per the specification in #30707 

_WARNING_: These are not currently hooked up to the discovery service or the token it creates.

Still TODO:
- [x] delete tokens
- [x] merge with #35144 and adopt it's testing approach
- [x] determine if we want wholesale json output & templating like kubectl (we do not have an API object with the data we want here) may require a bit of plumbing.
- [x] allow specifying a token duration on the CLI
- [x] allow configuring the default token duration
- [x] hook up the initial token created during init

Sample output:

```
(root@centos1 ~) $ kubeadm token create
Running pre-flight checks
<cmd/token> Token secret created: f6dc69.c43e491752c4a0fd
(root@centos1 ~) $ kubeadm token create
Running pre-flight checks
<cmd/token> Token secret created: 8fad2f.e7b78c8a5f7c7b9a
(root@centos1 ~) $ kubeadm token list  
Running pre-flight checks
ID        TOKEN                     EXPIRATION
44d805    44d805.a4e78b6cf6435e33   23h
4f65bb    4f65bb.d006a3c7a0e428c9   23h
6a086e    6a086e.2ff99f0823236b5b   23h
8fad2f    8fad2f.e7b78c8a5f7c7b9a   23h
f6dc69    f6dc69.c43e491752c4a0fd   23h
f81653    f81653.9ab82a2926c7e985   23h
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35805)

<!-- Reviewable:end -->

**Release note**:
```release-note
Added kubeadm commands to manage bootstrap tokens and the duration they are valid for.
```

dgoodwin committed:
Implement kubeadm bootstrap token management.

Adds kubeadm subcommands to create, list, and delete bootstrap tokens.
Tokens can be created with a TTL duration, or 0 for tokens that will not
expire. The create command can also be used to specify your own token
(for use when bootstrapping masters and nodes in parallel), or update an
existing token's secret or ttl.

Marked "ex" for experimental for now as the boostrap controllers are not
yet hooked up in core.

dgoodwin commented:
CC @jbeda @kubernetes/sig-cluster-lifecycle

luxas commented:
@dgoodwin Really nice! This is important
Is there anything I can do to help? Reviewing, etc..?

I understand that this is wip, so I'm asking before I touch it :)

pires commented on cmd/kubeadm/app/master/apiclient.go:
```diff
@@ -36,14 +36,14 @@ import (
 const apiCallRetryInterval = 500 * time.Millisecond
 
 func CreateClientAndWaitForAPI(adminConfig *clientcmdapi.Config) (*clientset.Clientset, error) {
+
```
Empty line.

pires commented on cmd/kubeadm/app/tokens/tokens.go:
```diff
@@ -0,0 +1,19 @@
+
+package tokens
+
+import ()
```
Please, remove empty import.

pires reviewed (changes requested):
Few nits. Also no tests or is it in-flight?

luxas commented:
@pires This is in my understanding a really rough first draft; see the TODO-list in the description

pires commented:
@luxas you're right, I should've done that instead of jumping to a code review.

@dgoodwin @jbeda probably should split this PR into several smaller ones? It would make it easier to review/merge. You could create an umbrella issue for keeping track of what's been delivered (as in merged).

dgoodwin commented:
@pires splitting the overall discovery implementation up into multiple PRs makes sense to me, this one is quite independent of the discovery API and auth work, hopefully we can push those all in separately. If you meant splitting *this* pr up let me know what you have in mind, it might slow things down as they would need sequential merges. (or if you meant merging this soon then more PRs for todos that might work fine)

@luxas @pires what do you guys think about the -o json and templating support? It involves a bit of wiring that looked a little nasty, but we could try if you feel it's important. It's a little strange here because we don't have an API type for this, the secret itself is unusable for the user, it has to be interpreted and displayed differently. I'm not sure if we can just make a type for that purpose and get away with it, most likely we could. (giving it a fake "kind" perhaps) All depends if folks think it's important though, we don't do that anywhere else in kubeadm and maybe we shouldn't bother.

pires commented:
> If you meant splitting this pr up let me know what you have in mind, it might slow things down as they would need sequential merges. (or if you meant merging this soon then more PRs for todos that might work fine)

Yeah, I meant to have those PRs merged quickly, so they need to be easy (fast) to review and merge. If it's a big PR, I'm afraid it may become a burden for you (rebase hell) and the reviewers.

Re: templating, I don't have an opinion, sorry. However, AFAIK `kubeadm` has its own API objects, alright (`MasterConfiguration`, `NodeConfiguration` and `ClusterInfo`).

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,206 @@
+	}
+	createCmd.PersistentFlags().StringVar(&tokenDuration,
+		"duration", "8h", "Duration the token should be valid")
+	tokenCmd.AddCommand(createCmd)
```
I guess we should have an option to specify the token that should be created.
I think something like this is fine
```
# kubeadm create token abcdef.1234567890123456 --duration 8h
abcdef.1234567890123456
# kubeadm create token --duration 8h
fedcba.6543210987654321
```

dgoodwin replied:
Right now it's in as kubeadm token create --token = MY.TOKEN. Let me know if you'd prefer the argless route.

luxas replied:
I think I prefer the argless route, but we should defer that decision until you've proceeded more on this PR and until others are taking a look at it

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,206 @@
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	createCmd.PersistentFlags().StringVar(&tokenDuration,
```
Can this be a duration instead of a string?

dgoodwin replied:
Wow it can, cool.

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,206 @@
+	return data
+}
+
+func createAPIClient() (*clientset.Clientset, error) {
```
Don't we have some code like this already in `apiclient.go`? If not, we should refactor out this code into a helper

dgoodwin replied:
It was based on that but quite different, however on a second pass I was able to merge the two and this is now a shared method.

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,206 @@
+	if _, err := client.Secrets(api.NamespaceSystem).Create(secret); err != nil {
+		return fmt.Errorf("<cmd/token> failed to bootstrap token [%v]", err)
+	}
+	fmt.Printf("<cmd/token> Token secret created: %s\n", tokenSecret.GivenToken)
```
I think we should just print the token here

luxas reviewed (commented):
Quick pass, some design comments

luxas commented on cmd/kubeadm/app/util/tokens.go:
```diff
@@ -60,6 +60,25 @@ func GenerateToken(s *kubeadmapi.Secrets) error {
 
+func GenerateTokenIfNeeded(s *kubeadmapi.Secrets) error {
+	ok, err := UseGivenTokenIfValid(s)
+	// TODO(phase1+) @krousey: I know it won't happen with the way it is currently implemented, but this doesn't handle case where ok is true and err is non-nil.
```
can you fix this at the same time :)?

dgoodwin replied:
Of course, done!

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+		Type: api.SecretTypeBootstrapToken,
+		Data: encodeTokenSecretData(tokenSecret, tokenDuration),
+	}
+	if _, err := client.Secrets(api.NamespaceSystem).Create(secret); err != nil {
```
We should make this an `upsert` where we update or create/insert.

Do this:
1. Run in a loop with a certain number of retries in case something flakes out.  Something like 5.
2. Try to get the secret.  
  1. If it is there, then replace the Data array and update.
  2. If it doesn't exist (`IsNotFound`) then create it and set it
3. If you get an error then retry from the top.  Errors might be `IsNotFound` (if you are doing an update and it was deleted between the get/update) or `IsConflict` (if you raced someone to do an update and lost).


See `pkg/api/errors/errors.go`

dgoodwin replied:
This should now be corrected.

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+	t := time.Now()
+	t = t.Add(duration)
+	data["expiration"] = []byte(t.Format(time.RFC3339))
+
```
`data["usage-bootstrap-signing"] = []byte("true")`

dgoodwin replied:
Added.

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+		return fmt.Errorf("<cmd/token> failed to list bootstrap tokens [%v]", err)
+	}
+
+	w := tabwriter.NewWriter(os.Stdout, 10, 4, 3, ' ', 0)
```
Should this go to "out"?

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+	}
+	w.Flush()
+
+	return nil
```
Eventually folks are going to ask for ways to script this and we'll find ourselves reimplementing all of the output modes that `kubectl` has.  This is good for now but won't last.

Also -- most users see the key as `123456.1234567890123456`.  I'm not sure we need to break it down to id and token (I called that part "secret").  I'd just name the column "token" and have the dotted form.

dgoodwin replied:
I implemented delete via the token ID, so my goal was to give the user two easy to copy parts, ID and full token which contains both parts with the '.'.

Are you suggesting just columns "token" and "expiration"? If so I could perhaps have "delete" handle both forms, and automatically dissect it if given the whole token.

jbeda replied:
Ah -- gotcha.  So the columns are `ID`, `ID.Secret`, `TTL`.  I misread it before and thought the second column was just the secret.  This looks good to me.

dgoodwin replied:
Sounds good, done.

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+	for _, secret := range results.Items {
+		tokenId := secret.Data["token-id"]
+		token := fmt.Sprintf("%s.%s", tokenId, secret.Data["token-secret"])
+		expireTime, err := time.Parse(time.RFC3339, string(secret.Data["expiration"]))
```
Handle the case where this doesn't exist and it doesn't expire.

dgoodwin replied:
Done!

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+	w := tabwriter.NewWriter(os.Stdout, 10, 4, 3, ' ', 0)
+	fmt.Fprintln(w, "ID\tTOKEN\tEXPIRATION")
+	for _, secret := range results.Items {
+		tokenId := secret.Data["token-id"]
```
there is no guarantee that these secrets are well formed.  Use the tuple return form (`tokenId, ok := secret.Data["token-id"]` and check `ok`.

If keys are malformed, I'd output a warning but continue on.  There will be some good keys that folks can use.   That means you probably want to send stderr in too.

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	tokenSecretName := fmt.Sprintf("%s%s", bootstrapTokenSecretPrefix, tokenId)
+	if err := client.Secrets(api.NamespaceSystem).Delete(tokenSecretName, nil); err != nil {
```
You may get `IsNotFound` here and that is okay.  That means that it was deleted while the user was doing something else -- perhaps expired?  I'd output something like "<cmd/token> bootstrap token already deleted: <tokenId>".  This shouldn't return an error.

jbeda commented on cmd/kubeadm/app/master/apiclient.go:
```diff
@@ -35,22 +35,35 @@ import (
 	).ClientConfig()
 	if err != nil {
-		return nil, fmt.Errorf("<master/apiclient> failed to create API client configuration [%v]", err)
+		return nil, fmt.Errorf("<cmd/token> failed to create API client configuration [%v]", err)
```
This isn't a token thing, necessarily.  This class will get used lots of places.

We should either pass the "phase" in or prepend the phase as we dump the error.

dgoodwin replied:
These were intended to just be debug info identifying what module/file the message was from, IIRC they might actually be short lived and going away in future. This one was a bad copy paste, it should be master/apiclient still, propose I fix it to master/apiclient and we'll sort that out with output cleanup in the future. It the method gets moved to a more generic module we can update prefix then as well.

jbeda commented:
Very cool!  No huge changes requested.  Biggest thing is around API server errors.

Do we want a release note here?  If we are releasing kubeadm with the rest of kubernetes we probably do.

dgoodwin commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+		},
+	}
+	createCmd.PersistentFlags().DurationVar(&tokenDuration,
+		"duration", defaultTokenDuration, "Duration the token should be valid")
```
TTL sounds good. 

Mentioning no TTL feels a little worrisome, do you feel that's important to expose if users want to opt into it? Or could we get away with just not offering an unlimited bootstrap token.

jbeda replied:
I'm cool with leaving "0 == unlimited" undocumented.  It is bad practice but I could see folks wanting to do it for test clusters.  We should make sure we can display unlimited tokens though in "list".

dgoodwin commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+	if _, err := client.Secrets(api.NamespaceSystem).Create(secret); err != nil {
+		return fmt.Errorf("<cmd/token> failed to bootstrap token [%v]", err)
+	}
+	fmt.Println(tokenSecret.GivenToken)
```
Seeing below I may have misunderstood.

jbeda commented:
@dgoodwin -- This is really shaping up.  But we are going to want to be careful how we stage this.  Specifically, we shouldn't make this the default for users until we get the server side stuff done and these tokens work for both discovery and TLS and are the only thing we use.  Otherwise we might just confuse users.

Perhaps put this under a "devel" or "experimental" node in the command structure so that for now it is `kubeadm devel token list`?

dgoodwin commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,183 @@
+	tokenSecret := &kubeadmapi.Secrets{}
+	createCmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create discovery tokens on the server.",
```
Agree, done!

dgoodwin commented:
We use "ex" in some openshift commands for experimental, how does that sound? (will add it this way for now as it's nice and snappy, if anyone feels strongly it should be elsewhere let me know)

dgoodwin commented:
I think I've addressed everything mentioned above. Separate commits for re-reviewability.

jbeda commented on cmd/kubeadm/app/util/tokens.go:
```diff
@@ -103,18 +105,45 @@ func UseGivenTokenIfValid(s *kubeadmapi.Secrets) (bool, error) {
+// UpdateOrCreateToken attempts to update a token with the given ID, or create if it does
+// not already exist.
+func UpdateOrCreateToken(client *clientset.Clientset, tokenSecret *kubeadmapi.Secrets, tokenDuration time.Duration) error {
+	secretName := fmt.Sprintf("%s%s", BootstrapTokenSecretPrefix, tokenSecret.TokenID)
```
The only "hole" here is if someone creates a bootstrap token from some other client with a different name but the same ID.  The "right" way to do this would be to list all the tokens and index those by ID.  Then you could update a token without making assumptions about the secret name.

But that probably isn't worth the trouble at this point.  As things stand the server impl will throw a warning and do its best.  Folks can always use a `kubectl delete secret` command to clean this up if need be.

jbeda commented:
Looking great!  Squash and I'll LGTM.

jbeda commented:
Also, the bazel verification is failing...

dgoodwin commented:
I rebased on top of @pipejakob 's work for the generate command, I kept it as a separate command, I briefly considered a create --print-only option but then we need to start watching for combinations with --ttl and --token, seemed best to just keep it separate for now.

@pipejakob please take a quick look at let me know if I botched anything in the merge.

Squashed and signed otherwise, should be ready to go.

dgoodwin commented:
Fixed another verify failure.

jbeda commented:
It looks good -- let me know when you want me to LGTM.

dgoodwin commented:
Just caught an issue this morning where the token command was also available top level as well as below "ex", removed now.

dgoodwin commented:
Ready when you are @jbeda

luxas commented:
@k8s-bot unit test this

luxas commented:
@k8s-bot cvm gke e2e test this

luxas commented:
LGTM, also it looked good to @jbeda earlier.

luxas commented:
@dgoodwin This unit failure looks real. PTAL

dgoodwin commented:
Thanks @luxas, updated the tests to "kubeadm ex token".

luxas commented:
@dgoodwin You can now rebase this and we'll get it merged :+1:

dgoodwin commented:
@k8s-bot test this

dgoodwin commented:
@k8s-bot test this

dgoodwin commented:
@luxas waiting on builds but I think this is ready now, big rebase conflict yesterday and then got bazel fixed this morning.

dgoodwin commented:
@k8s-bot cvm gke e2e test this

luxas commented:
@jbeda @mikedanese PTAL once more

luxas reviewed (approved):
Haven't had a chance to manually test this yet unfortunately, but it looks good to me.

I don not want to defer this PR anymore, let's just get it in and move on! We'll have lots of improvement possibilities to actually hook this flow up with something once @jbeda's PR is merged.

@jbeda @mikedanese Can I apply the label?
Does the rebase stuff look good?

jbeda commented:
It LGTM but I think the only open question is `beta` in the secret type.   @mikedanese -- thoughts there?

luxas commented:
@jbeda You can update it in your PR anyway, since you will have to rebase yours in any case :smile:

jbeda commented:
/lgtm

Final comment -- we should probably have a release note on this.  Can you add one to the description and reset the labels (or ping me to tweak them).

dgoodwin commented:
@jbeda I added a release note block in the description, I don't think I can change labels myself however. Let me know if there's anything else I need to do, thanks!

jbeda commented:
Fixed.  Thanks!

luxas commented:
@jbeda Just a side note, I'm absolutely not opposed to adding release-note labels for kubeadm PRs, but we have not done it with any PRs so far I think, at least not many.

I've aggregated the relnotes manually instead so far.
I'm not opposed to starting to add release-note labels for kubeadm now in time for v1.6, but since we're gonna break out soon or at least in a reasonably short timeframe, I'm not sure it's worth it to start a pattern that we know we're gonna break

luxas commented:
Hahah, seems like I repeated myself there a bit (writing half a comment - doing something else - writing the other part - publish :smile:)
